enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

```diff
@@ -0,0 +1,298 @@
+that cluster administrators will not want, instead of access to just the specific piece of content on the host file system.
+- Host injection via the MCO requires a node reboot for the updates to take effect.  That is generally deemed 
+undesirable for this feature.  
+- The MCO API has not been intuitive for the type of users to date who are interested in introducing and then consuming
+small collections of files on a host.
+
+This proposal will describe a [CSI driver](https://github.com/container-storage-interface/spec/blob/master/spec.md) based
+means of OpenShift cluster administrators the means of injecting K8S Secrets onto each host/node in an OpenShift cluster, 
+and then allow owners/authors/users of Pods to opt into the consumption of the content of those Secrets during their 
+Pod's execution.  The means of opting in will entail setting an annotation or label on an object vs. explicitly
+defining the volume related API objects or volume plug points into the Pod.
+
+This proposal was broken off from [a proposal on injecting RHEL entitlements/subscriptions](https://github.com/openshift/enhancements/pull/214)
+**EDITORIAL NOTE:** when PR 214 merges the reference link will be adjusted to point to the merged document hosted on
+https://github.com/openshift/enhancements.
+
+
+## Motivation
+
+Simplify overall usability for injecting and consuming Secret content onto the OpenShift hosts/nodes for use in user
+Pods.
+
+### Goals
+
+Provide alternatives to users specifying privileged Pods with HostPath volume.
+
+Provide alternatives to cluster administrators for using the MCO for injecting content onto nodes.
+
+### Non-Goals
+
+1. This proposal assumes the sensitive content encapsulated in the K8S Secret is valid and correct for the user's purposes.
+No validation of the content for specific usage within a Pod will be done.  That said, this proposal does provide 
+for watching for updates to the K8S Secrets being stored on the host/node by the associated CSI Driver of this proposal.
+So if the cluster administrator realized on their own they have provided bad data, and then corrects it, this proposal
+will watch for, discover, and propagate the update to the host/node filesystem.
+
+2. Protecting this content from being viewed/copied by users. If it's available for the Pod to use, the user who can 
+create a Pod can also read / view / exfiltrate the Secret contents. It is technically impossible to avoid that.
+
+3. 4.x customer feedback has also revealed a desire for namespace level granularity when injecting/consuming
+K8S Secret information available on the underlying host/node.  While this proposal will articulate some thought on 
+how to do this, ultimately a solution based on a given Secret utilized across multiple namespaces is deemed "meets
+minimum" for this proposal, and per namespace secret injection is out of scope.   
```

Comment:
arguably this proposal(specific to the CSI driver) doesn't concern itself with whether the content is "global" or user namespaced.

in implementation terms, there's no difference.  the CSI driver is going to pull content from some namespace (whether a "special global one" controlled by an admin, or a "user controlled/user specific one").  Where it gets the content from is immaterial to the CSI driver and will be determined entirely by the PVC that is created and the config params it includes.

where the global vs user specific concerns will manifest is in the admission controller behavior:  how flexible the pod annotation is in terms of specifying where the content comes from, whether we auto-inject the PVC/volumemount for the global content always, etc.

edit: reading your comments below on namespace support, i may be operating under a misconception of how this gets wired, but let's chat on slack and you can educate me.  It seems like the PVC is going to have to at least have metadata that indicates what secret it wants.  Also specifying what namespace that secret comes from doesn't seem like a big addition, but it sounds like there may be complexities on the backend PV creation or secret watching.

Reply:
Yeah I also have an open question about if/how PVC metadata might get propagated to the CSI spec'ed methods/types.

But sure we can chat via slack / bluejeans during offices hours as needed.

Reply:
@bparees I believe https://kubernetes-csi.github.io/docs/volume-datasources.html holds the answer for how we can propagate the namespace information

A PVC datasource translates to 

```
type TypedLocalObjectReference struct {
	// APIGroup is the group for the resource being referenced.
	// If APIGroup is not specified, the specified Kind must be in the core API group.
	// For any other third-party types, APIGroup is required.
	// +optional
	APIGroup *string `json:"apiGroup" protobuf:"bytes,1,opt,name=apiGroup"`
	// Kind is the type of resource being referenced
	Kind string `json:"kind" protobuf:"bytes,2,opt,name=kind"`
	// Name is the name of resource being referenced
	Name string `json:"name" protobuf:"bytes,3,opt,name=name"`
}
```

We should be able to specify the namespace with that.

When I cycle off bugs and back to active proposal mgmt, I'll update the proposal with this and strike out the open question pertaining to this.

Reply:
Fair enough / good point.

As per namespace is out of scope, I'm going to
a) claim more than due diligence done with it, in at least exploring options through the CSI flows
b) remove this section entirely
c) leave the conversation thread unresolved so that it is easier for folks to discover what transpired and where we landed.